### PR TITLE
fix(Transfomers): call `.toJSON` in `toSnakeCase`

### LINKED ISSUE #8742: Non-recursive InteractionResponses#showModal body transformation
### Which package is this bug report for?

discord.js

### Issue description

Basically, the typings do allow me to pass `ModalComponentData` into showModal method, which itself takes `JSONEncodable<APIActionRowComponent<APIModalActionRowComponent>> | ActionRowData<ModalActionRowComponentData>`, **though doesn't handle toJSON method**.

As turns out, in `InteractionResponses#showModal` for `body.data` transformation, this code does the thing:
```js
data: isJSONEncodable(modal) ? modal.toJSON() : this.client.options.jsonTransformer(modal)
```

So if you pass non-`JSONEncodable` modal data, it doesn't have `toJSON` method in it, which makes it to use default jsonTransformer - `toSnakeCase`, which doesn't use `toJSON` method in the child properties.

Would be good to either forbid to use `JSONEncodable` components in non-`JSONEncodable` modal data, or fix it in `InteractionResponses.showModal` method

### Code sample

```typescript
import {
  type ButtonInteraction,
  ActionRowBuilder,
  ComponentType,
  TextInputBuilder
} from "discord.js";

/* Just in case */
function showModal(interaction: ButtonInteraction) {
  interaction.showModal({
    customId: "whatever",
    title: "Example Modal",
    components: [
      new ActionRowBuilder({
        /* ... - doesn't really matter */
      })
    ]
  });

  /* second case */
  interaction.showModal({
    customId: "second",
    title: "Another Example Modal",
    components: [
      {
        type: ComponentType.ActionRow,
        components: [
          new TextInputComponent({
            /* ... */
          })
        ]
      }
    ]
  });
}
```

### Package version

14.6.0

### Node.js version

18.9.0

### Operating system

win32

### Priority this issue should have

Low (slightly annoying)

### Which partials do you have configured?

No Partials

### Which gateway intents are you subscribing to?

No Intents

### I have tested this issue on a development release

_No response_

### PR DESCRIPTION
**Pull request by RedGuy12**
**Please describe the changes this PR makes and why it should be merged:**
Makes `toSnakeCase` call `.toJSON` functions if they exist.
Fixes #8742
**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes

I could make calling `.toJSON` togglable via a boolean parameter if that is preferred to avoid breaking changes.

### CODE REVIEW

File: packages/discord.js/src/util/Transformers.js [hunk 1]
```diff
@@ -9,6 +10,7 @@ const snakeCase = require('lodash.snakecase');
  */
 function toSnakeCase(obj) {
   if (typeof obj !== 'object' || !obj) return obj;
+  if (isJSONEncodable(obj)) return obj.toJSON();
```
Comment (Jiralite) [inline packages/discord.js/src/util/Transformers.js:13]: I feel like this check should be done beneath line 14 where it checks `instanceof Date` as that I believe is JSON-encodeable, no?

Comment (RedGuy12) [reply packages/discord.js/src/util/Transformers.js:13]: I was not aware of that. Fixed.

**Review by kyranet — APPROVED** [review]
This makes the bold assumption that any JSON encodable will always yield snake_case properties, but that's (hopefully) almost always the case here, and for the places we use `toSnakeCase` for, it would be wrong to use JSON encodable classes that don't yield snake_case properties.

Either way, this seems to be triggered when using builders inside plain objects, which makes sense.

**Comment by RedGuy12** [comment]
Would it be better if I called `toSnakeCase` on `toJSON`'s return value?

**Comment by suneettipirneni** [comment]
> Would it be better if I called `toSnakeCase` on `toJSON`'s return value?

I would personally prefer this. 

Splitting functions by specific behaviors yields more flexible predictable code. `toSnakeCase` should only be concerned with transforming the input (as-is) to snake_case json, it shouldn't do anything else in addition to that.

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**